noir-projects/noir-contracts/contracts/nft_contract/src/test/transfer_to_private.nr

```diff
@@ -0,0 +1,92 @@
+}
+
+#[test]
+unconstrained fn transfer_to_private_to_a_different_account() {
+    // Setup without account contracts. We are not using authwits here, so dummy accounts are enough
+    let (env, nft_contract_address, sender, recipient, token_id) = utils::setup_and_mint(/* with_account_contracts */ false);
+
+    let note_randomness = unsafe_rand();
+    let transient_storage_slot_randomness = unsafe_rand();
+    // Sender will be the msg_sender/transfer_preparer in prepare_transfer_to_private
+    let transfer_preparer_storage_slot_commitment = pedersen_hash(
+        [sender.to_field(), transient_storage_slot_randomness],
+        NFT::TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+    );
+
+    // We prepare the transfer
+    let prepare_transfer_to_private_call_interface = NFT::at(nft_contract_address).prepare_transfer_to_private(
+        sender,
+        recipient,
+        note_randomness,
+        transient_storage_slot_randomness
+    );
+    env.call_private_void(prepare_transfer_to_private_call_interface);
+
+    // Finalize the transfer of the NFT
+    let finalize_transfer_to_private_call_interface = NFT::at(nft_contract_address).finalize_transfer_to_private(token_id, transfer_preparer_storage_slot_commitment);
+    env.call_public(finalize_transfer_to_private_call_interface);
+
+    // Store the finalized note in the cache
+    let mut context = env.private();
+    let recipient_npk_m_hash = get_current_public_keys(&mut context, recipient).npk_m.hash();
+    let private_nfts_recipient_slot = derive_storage_slot_in_map(NFT::storage().private_nfts.slot, recipient);
+
+    env.store_note_in_cache(
+        &mut NFTNote { token_id, npk_m_hash: recipient_npk_m_hash, randomness: note_randomness, header: NoteHeader::empty() },
+        private_nfts_recipient_slot,
+        nft_contract_address
+    );
+
+    // Recipient should have the note in their private nfts
+    utils::assert_owns_private_nft(nft_contract_address, recipient, token_id);
+
+    // Since the NFT got transferred to private public owner should be zero address
+    utils::assert_owns_public_nft(env, nft_contract_address, AztecAddress::zero(), token_id);
+}
+
+#[test(should_fail_with="transfer not prepared")]
+unconstrained fn transfer_to_private_to_self_transfer_not_prepared() {
+    // Setup without account contracts. We are not using authwits here, so dummy accounts are enough
+    let (env, nft_contract_address, _, _, token_id) = utils::setup_and_mint(/* with_account_contracts */ false);
+
+    // Transfer was not prepared so we can use random value for the commitment
+    let transfer_preparer_storage_slot_commitment = unsafe_rand();
+
+    // Try finalizing the transfer without preparing it
+    let finalize_transfer_to_private_call_interface = NFT::at(nft_contract_address).finalize_transfer_to_private(token_id, transfer_preparer_storage_slot_commitment);
+    env.call_public(finalize_transfer_to_private_call_interface);
+}
```

Comment:
We could also have a test in which we did prepare the transfer, but it was done for another account.

Reply:
Good idea. Addressed in https://github.com/AztecProtocol/aztec-packages/pull/8129/commits/d6dbf5dc170237791b040b0263892366e52f7b05